README.md
# api.gw2armory.com

[![Build Status](https://travis-ci.org/madou/armory-services.svg?branch=master)](https://travis-ci.org/madou/armory-services) [![Dependencies](https://david-dm.org/madou/armory-react.svg)](https://david-dm.org/madou/armory-services)
[![Dev Dependencies](https://david-dm.org/madou/armory-react/dev-status.svg)](https://david-dm.org/madou/armory-services?type=dev) 
[![codecov](https://codecov.io/gh/madou/armory-services/branch/master/graph/badge.svg)](https://codecov.io/gh/madou/armory-services)
[![Discord](https://img.shields.io/badge/discord-GW2Armory-blue.svg)](https://discord.gg/3BRbV7b)
[![Patreon](https://img.shields.io/badge/patreon-Become%20a%20Patreon-green.svg)](https://www.patreon.com/bePatron?u=5546924)

## Usage

### Quick start

tl;dr -> develop with `npm run tdd` and `npm run tdd:int`, use `npm run dev` to run local environment.

### User image uploads

To have image uploads working locally you'll need a valid aws key/secret pair. Run it like so:

```bash
IMAGE_UPLOAD_ACCESS_KEY_ID=ACCESSKEYHERE IMAGE_UPLOAD_SECRET_ACCESS_KEY=SECRETKEYHERE npm run dev
```

### Email notifications

To have email notification working locally you'll need a valid aws key/secret pair. Run it like so:

```bash
SES_ACCESS_KEY_ID=ACCESSKEYHERE SES_SECRET_ACCESS_KEY=SECRETKEYHERE npm run dev
```

### Database migrations

Ensure any database migration is backwards compatible with a previous version of the api. This will keep deployments simple and clean.

```bash
ENV={ENV} npm run migrate
```

#### Migration test runs

##### Prepare step

Starts a database container and exposes it to 127.0.0.1:3306

```bash
npm run mtr-prepare
```

##### Run step

Runs the migration against the test database. Make sure to have your migrations written in `src/migration/scripts`, with the db models in their origin (pre-migrated) state in `src/lib/models`.

```bash
npm run mtr
```

##### Revert step

Run after running test migration to roll back changes.

```bash
npm run mtr-revert
```

## Deployments

Deployments to production are triggered by tagged commits, tag with `npm version major|minor|patch`. Deployments to test are done adhoc and are _almost_ manual.

## Pull Requests

Like to contribute? Look at the [issues](https://github.com/madou/armory-services/issues) tab or contact me on [reddit](https://www.reddit.com/r/gw2armory) or [twitter](https://twitter.com/itsmadou) to find something you'd like to work on, then make a pull request against the `master` branch.

© 2015-present gw2armory.com


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][]

### Fixed

- Calling out to gw2 api with non-existent token

## [3.10.0][] - 2018-10-28

### Added

- Saving every standing record for historical lookup

## [3.9.6][] - 2018-10-27

### Fixed

- Pvp leaderboard fetching timing out

## [3.9.5][] - 2018-06-28

### Fixed

- Docker login script

## [3.9.4][] - 2018-06-22

### Fixed

- Item states not being calculated correctly from breaking change in ANet API

## [3.9.3][] - 2017-10-11

### Fixed

- Item stats being calculated incorrectly

## [3.9.2][] - 2017-10-09

### Fixed

- Single serving itemstats resource

## [3.9.1][] - 2017-10-09

### Fixed

- Bulk request mode for itemstats to handle errors

### Changed

- Dependencies to be the latest versions
- Sequelize pool options

## [3.9.0][] - 2017-10-07

### Added

- Bulk request mode for itemstats

## [3.8.2][] - 2017-10-06

### Changed

- Use custom version of gw2itemstats

## [3.8.1][] - 2017-10-06

### Fixed

- Calclated item stats not setting lang

## [3.8.0][] - 2017-10-06

### Added

- Calculated item stats resource

### Fixed

- Logging serialization

## [3.7.2][] - 2017-09-27

### Changed

- User access to be array based output

## [3.7.1][] - 2017-09-22

### Changed

- Use `serialize-error` to parse errors to logger

## [3.7.0][] - 2017-09-21

### Added

- Missing user api proxy routes

## [3.6.0][] - 2017-09-07

### Removed

- CORS origin restrictions

## [3.5.0][] - 2017-09-06

### Added

- Season data to pvp leaderboards endpoint

## [3.4.5][] - 2017-09-02

### Fixed

- Response not being sent back for unclaimed users
- Sequelize request not bringing back associated models

## [3.4.4][] - 2017-08-29

### Changed

- Updated sequelize
- Updated restify
- Other dependencies

## [3.4.3][] - 2017-08-27

### Fixed

- Humanizes specific errors to not get crazy objects inside slack

## [3.4.2][] - 2017-08-16

### Changed

- Use `circular-json` dep instead of `JSON.stringify` for error logging

## [3.4.1][] - 2017-08-15

### Changed

- Upgrades deps

## [3.4.0][] - 2017-08-06

### Added

- Ability to reclaim api token when removed from arenanet but not the armory

## [3.3.1][] - 2017-06-05

### Changed

- EB environment to single instance to reduce costs

## [3.3.0][] - 2017-06-04

### Added

- New privacy system for all resources
- New data for users

## [3.2.6][] - 2017-05-19

### Fixed

- Slack messaging not working

## [3.2.5][] - 2017-05-18

### Fixed

- Sitemap index not generating all pages

## [3.2.4][] - 2017-05-18

### Fixed

- Xml parser wrecking it

## [3.2.3][] - 2017-05-18

### Changed

- Sitemap to be calculated in a performant manner

## [3.2.2][] - 2017-05-06

### Fixed

- Character and user authentication causing problems for authenticated users

## [3.2.1][] - 2017-04-30

### Fixed

- Api token migration script

## [3.2.0][] - 2017-04-30

### Fixed

- Pvp fetcher fetching when user doesn't have pvp permissions
- Guilds of the day to be cached for the day

### Changed

- Fetch logging to be more granular

### Added

- Api token invalidator
- Log when unhandled rejections/errors happen

## [3.1.0][] - 2017-04-28

### Added

- Try fetch capability to all resources

## [3.0.1][] - 2017-04-27

### Changed

- Gitter to slack logging

## [3.0.0][] - 2017-04-26

### Changed

- Babel configuration to use `babel-preset-env`
- Docker images to use node `7.8.0`
- Leaderboard pagination to match everything else
- Guild members to be orded by armory users first
- Gw2 api fetch to cache in memory for 5 minutes

### Removed

- Characters from base guild resource
- Characters from user resource
- Users from base guild resource

### Added

- guild/characters resource
- guild/users resource
- Characters/guilds of the day

### Fixed

- Guild resource to return 404 if not found

## [2.4.0] - 2017/04/05

### Added

- Character tabs to sitemap

## [2.3.0] - 2017/03/29

### Changed

- Deployment process to deploy to production with tagged commits

### Added

- Precommit hook
- Code coverage

## [2.2.8] - 2017/03/29

### Changed

- Fetch config to fetch as slow as possible for now

## [2.2.7] - 2017/03/23

### Changed

- Fetch to be enabled

## [2.2.6] - 2017/03/21

### Fixed

- `getLatestPvpSeason` from breaking half the site when GW2API is down

## [2.2.5] - 2017/03/20

### Fixed

- Users `me` resource from not catching exception and returning

## [2.2.4] - 2017/03/20

### Changed

- Turned off fetch temporarily while gw2 apis are down

## [2.2.3] - 2017/03/08

### Fixed

- User image uploads from not working

## [2.2.2] - 2017/03/07

### Added

- Corrects changelog

## [2.2.1] - 2017/03/01

### Removed

- CORS restrictions

## [2.2.0] - 2017/02/27

### Added

- Version to healthcheck endpoint

## [2.1.1] - 2017/02/27

### Removed

- Standing stats for wins/losses from being fetched (it's a combined wins/losses, not specific to the current season!). If https://github.com/arenanet/api-cdi/issues/452 is ever implemented, add this back.

## [2.1.0] - 2017/02/25

### Added

- Wins/losses to user resource

## [2.0.0] - 2017/02/25

### Added

- Wins/losses to pvp standings

### Removed

- Properties from leaderboard endpoints that weren't needed

## [1.7.1] - 2017/02/20

### Fixed

- Local environment from not being able to start up correctly

## [1.7.0] - 2017/02/16

### Added

- Advanced health check for beanstalk

## [1.6.3] - 2017/02/13

### Changed

- Gw2 api to filter out undefined items

## [1.6.2] - 2017/02/03

### Changed

- Sitemap code to be a little cleaner
- Travis ci to use yarn by the official means

## [1.6.1] - 2017/02/03

### Changed

- Sitemap to be paginated

## [1.6.0] - 2017/02/03

### Added

- Leaderboard pagination

### Fixed

- Leaderboard calculation from not clearing na/eu users who dropped off the ladder

go.sh
#!/bin/bash

APP_VERSION="1.0.0"
APP_NAME="application"
MYSQL_PASS="password"
MYSQL_USER="admin"
MYSQL_DB="armory"
DOCKER_IMAGE_PREFIX="gw2armory"
DOCKERHUB_PREFIX="madou/$DOCKER_IMAGE_PREFIX"

log() {
  printf "\n\n== $1 ==\n\n"
}

pause() {
  printf "Pausing for $1 seconds.."

  ((t = $1))

  while ((t > 0)); do
      printf "."
      sleep 1
      ((t -= 1))
  done

  log " Finished!"
}

clean() {
  log "Cleaning up exited containers and untagged images..."
  docker ps -a | grep Exit | cut -d ' ' -f 1 | xargs docker rm
  docker images | grep "<none>" | awk '{print $3}' | xargs docker rmi -f
}

remove() {
  log "Removing container $DOCKER_IMAGE_PREFIX-$1.."
  docker rm -f "$DOCKER_IMAGE_PREFIX-$1"
}

push() {
  docker push "$DOCKERHUB_PREFIX-$1"
}

run() {
  log "Running daemon $1:$2.."

  local TAG="$DOCKERHUB_PREFIX-$1:$APP_VERSION"
  local NAME="$DOCKER_IMAGE_PREFIX-$2"

  docker run \
    -d \
    $3 \
    --name $NAME \
    -e "ENV" \
    -e "IMAGE_UPLOAD_ACCESS_KEY_ID" \
    -e "IMAGE_UPLOAD_SECRET_ACCESS_KEY" \
    -e "SES_ACCESS_KEY_ID" \
    -e "SES_SECRET_ACCESS_KEY" \
    -e "SLACK_BOT_TOKEN" \
    $TAG \
    $4
}

exec() {
  log "Running $1.."

  local TAG="$DOCKERHUB_PREFIX-$APP_NAME:$APP_VERSION"
  local NAME="$DOCKER_IMAGE_PREFIX-$1"

  docker run \
    --rm \
    $3 \
    --name $NAME \
    $TAG \
    $2
}

build() {
  log "Building app image.."

  docker build \
    -t "$DOCKERHUB_PREFIX-$1:$APP_VERSION" \
    -t "$DOCKERHUB_PREFIX-$1:latest" \
    $2
}

kill() {
  remove db
  remove fetch
  remove api
}

runDb() {
  run db db "-e MYSQL_ROOT_PASSWORD=$MYSQL_PASS -e MYSQL_PASSWORD=$MYSQL_PASS -e MYSQL_USER=$MYSQL_USER -e MYSQL_DATABASE=$MYSQL_DB -p 3306:3306"

  pause 30
}

runFetch() {
  run $APP_NAME fetch "--link $DOCKER_IMAGE_PREFIX-db:db" "npm run fetch"
}

runApi() {
  run $APP_NAME api "-p 80:80 --link $DOCKER_IMAGE_PREFIX-db:db --link $DOCKER_IMAGE_PREFIX-fetch:fetch" "npm run api"
}

# Runs environment including database
dev() {
  remove db
  remove fetch
  remove api

  build db ./src/db
  build $APP_NAME .

  runDb
  runFetch
  runApi
}

setupMtr() {
  build db ./src/db
  remove db
  runDb
}

# Runs environment without database
start() {
  remove fetch
  remove api

  build $APP_NAME .

  runFetch
  runApi
}

case "$1" in
  build)
    build $APP_NAME .;;
  push)
    push $APP_NAME;;
  dev)
    dev;;
  start)
    start;;
  mtr)
    testmigrate;;
  db)
    setupMtr;;
  clean)
    clean;;
  kill)
    kill;;
  *)
    exit 1;;
esac
